### PR TITLE
Remove LLVM 14 support

### PR DESCRIPTION
**Pull request by aykevl**
This is a big change: apart from removing LLVM 14 it also removes typed pointer support (which was only fully supported in LLVM up to version 14). This removes about 200 lines of code, but more importantly removes a ton of special cases for LLVM 14.

This will need to be rebased once #3741 is merged.

### CODE REVIEW

**Comment by deadprogram** [comment]
@aykevl you can rebase this against `dev` now, if you please.

**Comment by aykevl** [comment]
Rebased.

**Comment by dgryski** [comment]
Looks like genuine CI failures:
```
/home/runner/work/tinygo/tinygo/targets/avrtiny.S:5:9: error: invalid operand for instruction
    clr __zero_reg__ ; this register is expected to be 0 by the C calling convention
        ^
/home/runner/work/tinygo/tinygo/targets/avrtiny.S:38:12: error: invalid operand for instruction
    st X+, __zero_reg__ ; *(x++) = 0
           ^
failed to run tool: clang
```

**Comment by aykevl** [comment]
Thanks, updated the PR and rebased to fix the Windows cache issue (hopefully it passes now).

File: targets/avrtiny.S [hunk 1]
```diff
@@ -1,4 +1,3 @@
-; TODO: remove these in LLVM 15
 #define __tmp_reg__ r16
 #define __zero_reg__ r17
```
Comment (aykevl) [inline targets/avrtiny.S:2]: I assumed these defines could be removed in LLVM 15. I was wrong, they are still needed (`__tmp_reg__` and `__zero_reg__` are only available in inline assembly). Hence why I just removed the comment.

**Review by dgryski — APPROVED** [review]
LGTM

**Pull request merged**